tensorflow/lite/experimental/shlo/ops/dot_general.cc

```diff
@@ -0,0 +1,441 @@
+              zero_point_value.begin(), zero_point_value.end(),
+              [](I4 value) { return value == static_cast<I4>(0); });
+        } else if (std::holds_alternative<absl::InlinedVector<int8_t, 8>>(
+                       zero_points)) {
+          auto zero_point_value =
+              std::get<absl::InlinedVector<int8_t, 8>>(zero_points);
+          return std::all_of(
+              zero_point_value.begin(), zero_point_value.end(),
+              [](int8_t value) { return value == static_cast<int8_t>(0); });
+        } else if (std::holds_alternative<absl::InlinedVector<int16_t, 8>>(
+                       zero_points)) {
+          auto zero_point_value =
+              std::get<absl::InlinedVector<int16_t, 8>>(zero_points);
+          return std::all_of(
+              zero_point_value.begin(), zero_point_value.end(),
+              [](int16_t value) { return value == static_cast<int16_t>(0); });
+        }
+        return false;
+      };
+      if (!check_zero_points_values_are_zero(
+              rhs.quantized_per_axis_element_type().ZeroPoints())) {
+        return absl::FailedPreconditionError(
+            "stablehlo.dot_general: The rhs per-axis should have zero points "
+            "as 0.");
+      }
+    } else if (rhs.IsPerAxisQuantized()) {
+      for (DimensionSize i = 0; i < rhsc_size; ++i) {
+        if (rhsc[i] ==
+                rhs.quantized_per_axis_element_type().QuantizedDimension() ||
+            rhsb[i] ==
+                rhs.quantized_per_axis_element_type().QuantizedDimension()) {
+          return absl::FailedPreconditionError(
+              "stablehlo.dot_general: If the rhs is per-axis quantized than "
+              "the quantization_dimensions of rhs should not be in "
+              "rhs_contracting_dimensions and rhs_batching_dimensions.");
+        }
+      }
+    }
+  }
+  return absl::OkStatus();
+}
+
+template <DataType storage_type>
+absl::Status EvaluateImpl(DotGeneralOp& op, const Tensor& lhs,
+                          const Tensor& rhs,
+                          const Tensor& lhs_batching_dimensions,
+                          const Tensor& rhs_batching_dimensions,
+                          const Tensor& lhs_contracting_dimensions,
+                          const Tensor& rhs_contracting_dimensions,
+                          Tensor& output) {
+  using StorageT = StorageType<storage_type>;
+  const StorageT* lhs_data = lhs.GetDataAs<storage_type>();
+  const StorageT* rhs_data = rhs.GetDataAs<storage_type>();
+  StorageT* output_data = output.GetDataAs<storage_type>();
+  const size_t lhs_size = lhs.NumElements();
+  const DimensionSize rhs_size = rhs.NumElements();
+  const DimensionSize output_size = output.NumElements();
+  const size_t lhs_rank = lhs.Rank();
+  const size_t rhs_rank = rhs.Rank();
+  const size_t output_rank = output.Rank();
+
+  const int32_t* lhs_batching_dimensions_data =
+      lhs_batching_dimensions.GetDataAs<DataType::kSI32>();
```

Comment:
`SI64`.  This type is currently missing. A fix is on its way.

Reply:
changed to SI64